src/main/scala/is/hail/expr/FunctionRegistry.scala

```diff
@@ -237,10 +226,109 @@ object FunctionRegistry {
           v = args(2).run(ec)();
 
           _ = (if (v == null)
-            fatal(s"Argument evaluated to missing in call to aggregator $name"));
+            fatal(s"Argument evaluated to missing in call to aggregator $name"))
+
+        ) yield aggregator.ctor(g, v.asInstanceOf[v])
 
-          aggregationResultThunk <- addAggregation(args(0), aggregator.ctor(g, v.asInstanceOf[v]));
+      case aggregator: GroupByAggregatorFun[t] =>
+        val Lambda(_, keyParam, keyBody) = args(1)
+        val TFunction(Seq(keyParamType), _) = argTypes(1)
 
+        val Lambda(_, downstreamParam, downstreamBody) = args(2)
+        val TFunction(Seq(downstreamParamType), _) = argTypes(2)
+
+        val kbc = keyBody.compile()
+
+        val a0type = args(0).`type`
+
+        for (
+          (downstreamBodyCompiled, downstreamAggregator) <- downstreamBody.downstreamAggregator();
+
+          ec <- ec();
+          (idx, localA) <- ecNewPosition();
+
+          bodyST = a0type match {
+            case tagg: TAggregable => tagg.symTab
+            case _ => ec.st
+          };
+
+          keyBodyFn = (for (
+            fb <- fb();
+            bindings = (bodyST.toSeq
+              .map { case (name, (i, typ)) =>
+                (name, typ, ret(Code.checkcast(fb.arg2.invoke[Int, AnyRef]("apply", i))(typ.scalaClassTag)))
+            } :+ ((keyParam, keyParamType, ret(Code.checkcast(fb.arg2.invoke[Int, AnyRef]("apply", idx))(keyParamType.scalaClassTag)))));
+            res <- bindRepInRaw(bindings)(kbc)
+          ) yield res).runWithDelayedValues(bodyST.toSeq.map { case (name, (_, typ)) => (name, typ) }, ec);
+
+          key = (x: Any) => {
+            localA(idx) = x
+            keyBodyFn(localA.asInstanceOf[mutable.ArrayBuffer[AnyRef]])
+          };
+
+          (downstreamParamId, localA2) <- ecNewPosition();
+
+          downstreamParamTypeFixed = {
+            val x = downstreamParamType.asInstanceOf[TAggregable].copy();
+            // FIXME filter out old bindings
```

Comment:
Very good points. I'm still happy to put this in in the meantime, though, but would like to understand a little better to anticipate edge cases.